python/ray/serve/_private/application_state.py

```diff
@@ -435,3 +623,25 @@ def update(self):
             record_extra_usage_tag(
                 TagKey.SERVE_NUM_APPS, str(len(self._application_states))
             )
+
+    def shutdown(self) -> None:
+        for app_state in self._application_states.values():
+            app_state.delete()
+
+    def _save_checkpoint_func(
+        self, *, writeahead_checkpoints: Optional[Dict[str, ApplicationTargetState]]
+    ) -> None:
+        """Write a checkpoint of all application states."""
+
+        application_state_info = {
+            app_name: app_state.get_checkpoint_data()
+            for app_name, app_state in self._application_states.items()
+        }
+
+        if writeahead_checkpoints is not None:
+            application_state_info.update(writeahead_checkpoints)
+
+        self._kv_store.put(
+            CHECKPOINT_KEY,
+            cloudpickle.dumps(application_state_info),
```

Comment:
Do you need to use `cloudpickle` to dump and load `application_state_info`? If you can use `pickle` instead, I'd recommend doing that instead. It's faster, and it prevents code from being serialized.